### PR TITLE
Monorepo update 6

### PR DESCRIPTION
**Pull request by ParthDesai**
## Description
This PR adds new features in the SDK.  There are two main additions:

1. `DestructorSet`: A data structure created from `DropCollection` and `AsyncDropFutures` data structure. This allows us to drop data structures/run sync and async destructors in exact order instead of rely on rust system.
2. `Domains v2 rewrite`: Domain part of sdk is rewritten to work with Domain v2. Since domain v2 starts in the phases, the design involves multiple async task co-ordinating and `DestructorSet` to try to shutdown cleanly.

The last 6 commits contains major changes. Commit wise changes are as follows: 
1. [Add Combined piece getter to prevent un-necessary requests to DSN](https://github.com/subspace/subspace-sdk/pull/205/commits/cd476f4ebbc99a7251e9c5a99f12021184da1cca): Prevents farmer to connect to its own shared DSN instance in order to get the piece. We do this by use combined piece getter that queries both Node and Farmer cache.
2. [Adding destructor to have deterministic order in which resources are …](https://github.com/subspace/subspace-sdk/pull/205/commits/a3e6c9ca98b10db7e7078cfce17cd8ffed85051a)
[a3e6c9c](https://github.com/subspace/subspace-sdk/pull/205/commits/a3e6c9ca98b10db7e7078cfce17cd8ffed85051a): This commit add support for `DestructorSet` data structure that will allow us to define a deterministic order in which resources are freed. This is done by maintaining the order of addition of sync destructor/async destructor and opaque object addition.
3. [rename destructors + adding tracing and much needed validation](https://github.com/subspace/subspace-sdk/pull/205/commits/d11abac0bbc256266cde0a49c04433e89b39c969): Minor commit that renames the `DestructorSet` data structure and adds validation and tracing inside it.
4. [add initial support for domains](https://github.com/subspace/subspace-sdk/pull/205/commits/dbe362e3074b3166a1520772887634784e676982): This commit adds initial support for Domains v2. 
5. [add concept of TaskOutput and handle cancellation in better way](https://github.com/subspace/subspace-sdk/pull/205/commits/93ae5b6787a6541163676f15addf595e3be281dc): Previously task cancellation and completion of task were not differentiated. In order to give better log output and be able to differentiate task completion and cancellation, we are using `TaskOutput` enum which can be `Ok` on completion and `Cancelled` on task cancellation.

Other notable changes from monorepo:
1. The Shared DSN Object has Role set as farmer.
2. The network keypair is node's keypair. As Network is initialized when we are building Node. (This change predates this PR but I am mentioning it just for completeness.)
3. Since DSN is shared, our `LocalRecordProvider` is combination of `FarmerPieceCache` and `NodePieceCache` and the we try to fetch the record from `FarmerPieceCache` first and fallback to `NodePieceCache`.

Third part of the larger PR: https://github.com/subspace/subspace-sdk/pull/201

### CODE REVIEW

**Review by shamil-gadelshin — CHANGES_REQUESTED** [review]
It makes sense in general. I tested L2 cache with `mini-farmer` example and `benchmark` example from monorepo - it seems working. I didn't review domains related code. 

General comments:
- KADEMLIA_PROVIDER_TTL_IN_SECS is obsolete and should be removed.
- "// TODO: check for piece cache to exit" seems obsolete

**Comment by shamil-gadelshin** [comment]
I would also set `kademlia_mode: Some(Mode::Server),` in the DSN configuration.

File: farmer/src/lib.rs [hunk 1]
```diff
@@ -525,43 +368,50 @@ impl Config {
                 node.rpc().clone(),
                 kzg.clone(),
                 // TODO: Consider introducing and using global in-memory segment commitments cache
-                parking_lot::Mutex::new(lru::LruCache::new(SEGMENT_COMMITMENTS_CACHE_SIZE)),
+                parking_lot::Mutex::new(lru::LruCache::new(RECORDS_ROOTS_CACHE_SIZE)),
```
[the remark this answers is not in the document]
Comment (ParthDesai) [reply farmer/src/lib.rs:371]: Hmm. But in the monorepo it is `RECORDS_ROOTS_CACHE_SIZE`.

[the remark this answers is not in the document]
Comment (nazar-pc) [reply farmer/src/lib.rs:371]: Should be renamed there as well, record roots is an old name

**Comment by ParthDesai** [comment]
> It makes sense in general. I tested L2 cache with `mini-farmer` example and `benchmark` example from monorepo - it seems working. I didn't review domains related code.
> 
> General comments:
> 
>     * KADEMLIA_PROVIDER_TTL_IN_SECS is obsolete and should be removed.
> 
>     * "// TODO: check for piece cache to exit" seems obsolete

Done.

**Review by shamil-gadelshin — APPROVED** [review]

**Review by vedhavyas — COMMENTED** [review]
Make sense. 

Not a blocker but I see a lot of repetitive code for domain here taken from monorepo. This might not be sustainable to update everytime. Why not pull it as crate from monorepo and use it here ?

File: node/src/chain_spec.rs [hunk 2]
```diff
@@ -328,13 +339,45 @@ fn subspace_genesis_config(
...
+
+                // Domain config, mainly for placeholder the concrete value TBD
+                owner_account_id: sudo_account.clone(),
+                domain_name: b"evm-domain".to_vec(),
+                max_block_size: MaxDomainBlockSize::get(),
+                max_block_weight: MaxDomainBlockWeight::get(),
+                bundle_slot_probability: (1, 1),
+                target_bundles_per_block: 10,
+                raw_genesis_config: raw_evm_domain_genesis_config,
+                // TODO: Configurable genesis operator signing key.
```
[the remark this answers is not in the document]
Comment (ParthDesai) [reply node/src/chain_spec.rs:375]: @vedhavyas This will be done in next PR. This PR is only till 11th August snapshot.

**Comment by ParthDesai** [comment]
> Make sense.
> 
> Not a blocker but I see a lot of repetitive code for domain here taken from monorepo. This might not be sustainable to update everytime. Why not pull it as crate from monorepo and use it here ?

Primary reason is: 
This logic is contained in `subspace-node` crate, which is high level crate with commands defined. If I use the crate, I will lose my ability to manage the tasks and extract the progress handles from it. On SDK side, I want to manage tasks and resource such that any `build` call (equivalent to `cli.run`) is not blocking and we will be able to close it cleanly. 

One way, we can improve this in future is we can move some part of domain instance starter out of the `subspace-node`  make it non-blocking. This way I can directly import and work with it to manage the tasks.

**Pull request merged**